src/feed.sol

```diff
@@ -0,0 +1,144 @@
+pragma solidity >=0.5.15;
+
+import "ds-note/note.sol";
+import "tinlake-auth/auth.sol";
+import "tinlake-math/interest.sol";
+import "./nftfeed.sol";
+import "./buckets.sol";
+
+contract Feed is BaseNFTFeed, Interest, Buckets {
+    // nftID => maturityDate
+    mapping (bytes32 => uint) public maturityDate;
+
+    // risk => recoveryRatePD
+    mapping (uint => uint) public recoveryRatePD;
+
+    // loan => futureValue
+    mapping (uint => uint) public futureValue;
+
+    WriteOff [2] public writeOffs;
+
+    struct WriteOff {
+        uint rateGroup;
+        // denominated in RAY (10^27)
+        uint percentage;
+    }
+
+    uint public discountRate;
+    uint public maxDays;
+
+    constructor (uint discountRate_, uint maxDays_) public {
+        discountRate = discountRate_;
+        maxDays = maxDays_;
+    }
+
+    function init() public {
+        super.init();
+        // gas optimized initialization of writeOffs
+        // write off are hardcoded in the contract instead of init function params
+
+        // risk group recoveryRatePD
+        recoveryRatePD[0] = ONE;
+        recoveryRatePD[1] = 90 * 10**25;
+
+        // 60% -> 40% write off
+        writeOffs[0] = WriteOff(91, 6 * 10**26);
+        // 80% -> 20% write off
+        writeOffs[1] = WriteOff(90, 8 * 10**26);
+    }
+
+    function uniqueDayTimestamp(uint timestamp) public pure returns (uint) {
+        return (1 days) * (timestamp/(1 days));
+    }
+
+    /// maturityDate is a unix timestamp
+    function file(bytes32 what, bytes32 nftID_, uint maturityDate_) public {
+        if (what == "maturityDate") {
+            maturityDate[nftID_] = uniqueDayTimestamp(maturityDate_);
+        } else { revert("unknown config parameter");}
+    }
+
+    /// Ceiling Implementation
+    function borrow(uint loan, uint amount) external auth {
+        uint normalizedDay = uniqueDayTimestamp(now);
+
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+
+        bytes32 nftID_ = nftID(loan);
+        uint maturityDate_ = maturityDate[nftID_];
+
+        // calculate future value FV
+        uint fv = rmul(rmul(rpow(pile.loanRates(loan),  safeSub(maturityDate_, normalizedDay), ONE), amount), recoveryRatePD[risk[nftID(loan)]]);
+
+        futureValue[loan] = safeAdd(futureValue[loan], fv);
+
+        if (buckets[maturityDate_].value == 0) {
+            addBucket(maturityDate_, fv);
+            return;
+        }
+
+        buckets[maturityDate_].value = safeAdd(buckets[maturityDate_].value, fv);
+    }
+
+    /// adds a new bucket to the linked-list
+    function repay(uint loan, uint amount) external auth {
+        uint maturityDate_ = maturityDate[nftID(loan)];
+
+        // todo reduce bucket with loan futureValue relative to amount
+        buckets[maturityDate_].value = safeSub(buckets[maturityDate_].value, amount);
+
```

Comment:
It is actually wrong that we are storing the FV. We actually need to store the relation in the mapping.

**Example:**

```
amount: 100
FV: 120

loan repay amount: 50
bucket reduction: ? 
```

**Calculation:**
amount/100 * 120=
50/100 * 120 = 60

Instead of storing the FV we should store principal/FV in the borrow method:
In our example: 
principal/FV = 100/120 = 1.2

In the repay method we only need to multiple `amount * 1.2`.

@ilinzweilin  do you agree?